notebooks/02_model/cornac_bpr_deep_dive.ipynb

```diff
@@ -0,0 +1,601 @@
+    "train, test = python_random_split(data, 0.75)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.2 Cornac Dataset\n",
+    "\n",
+    "To work with models implemented in Cornac, we need to construct an object from [Dataset](https://cornac.readthedocs.io/en/latest/data.html#module-cornac.data.dataset) class.\n",
+    "\n",
+    "Dataset Class in Cornac serves as the main object that the models will interact with.  In addition to data transformations, Dataset provides a bunch of useful iterators for looping through the data, as well as supporting different negative sampling techniques."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of users: 943\n",
+      "Number of items: 1642\n"
+     ]
+    }
+   ],
+   "source": [
+    "train_set = cornac.data.Dataset.from_uir(train.itertuples(index=False), seed=SEED)\n",
+    "\n",
+    "print('Number of users: {}'.format(train_set.num_users))\n",
+    "print('Number of items: {}'.format(train_set.num_items))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.3 Train the BPR model\n",
+    "\n",
+    "The BPR has a few important parameters that we need to consider:\n",
+    "\n",
+    "- `k`: controls the dimension of the latent space (i.e. the size of the vectors  $w_u$  and  $h_i$ ).\n",
+    "- `max_iter`: defines the number of iterations of the SGD procedure.\n",
+    "- `learning_rate`: controls the step size $\\alpha$ in the gradient update rules.\n",
+    "- `lambda_reg`: controls the L2-Regularization $\\lambda$ in the objective function.\n",
+    "\n",
+    "Note that different values of `k` and `max_iter` will affect the training time.\n",
+    "\n",
+    "We will here set `k` to 200, `max_iter` to 100, `learning_rate` to 0.01, and `lambda_reg` to 0.001. To train the model, we simply need to call the `fit()` method."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "bpr = cornac.models.BPR(\n",
+    "    k=NUM_FACTORS,\n",
+    "    max_iter=NUM_EPOCHS,\n",
+    "    learning_rate=0.01,\n",
+    "    lambda_reg=0.001,\n",
+    "    verbose=True,\n",
+    "    seed=SEED\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|██████████████████████████████████████████████████████████| 100/100 [00:07<00:00, 14.10it/s, correct=92.19%, skipped=9.38%]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Optimization finished!\n",
+      "Took 7.196771621704102 seconds for training.\n"
+     ]
+    }
+   ],
+   "source": [
+    "start_time = time.time()\n",
```

Comment:
we have a cool timer class: https://github.com/microsoft/recommenders/blob/master/reco_utils/common/timer.py

You could do something like:
```
with Timer() as t:
    bpr.fit(train_set)
print("Took {} seconds".format(t))
```

Reply:
updated!